project/Build.scala

```diff
@@ -712,27 +739,121 @@ object Build {
 
   lazy val linker: Project = (project in file("linker/jvm")).settings(
       commonLinkerSettings,
+      freezeVersionForToolsSettings,
       libraryDependencies ++= Seq(
           "com.google.javascript" % "closure-compiler" % "v20190513",
           "com.novocode" % "junit-interface" % "0.9" % "test"
       ) ++ (
           parallelCollectionsDependencies(scalaVersion.value)
       ),
+
+      resourceGenerators in Compile += Def.task {
+        val s = streams.value
+        val baseResourceDir = (resourceManaged in Compile).value
+        val resourceDir = baseResourceDir / "org/scalajs/linker/backend/emitter"
+
+        /* Awful hack to execute `products in (linkerPrivateLibrary, Compile)`
+         * as if we were in ++2.12.8, regardless of the current version of the
+         * linker and other projects.
+         */
+        //val privateLibProducts = {
+        //  val s = state.value
+        //  val relevantProjects = List(compiler, library, libraryAux, scalalib,
+        //      javalanglib, javalib, linkerPrivateLibrary)
+        //  val setVersionSettings = relevantProjects.map { p =>
+        //    scalaVersion in p := "2.12.8"
+        //  }
+        //  val newState =
+        //    appendWithSession(Project.extract(s), setVersionSettings, s)
+        //  val extracted = Project.extract(newState)
+        //  val (_, privateLibProducts) =
+        //    extracted.runTask(products in (linkerPrivateLibrary, Compile), newState)
+        //  privateLibProducts
+        //}
```

Comment:
@dwijnand you recommended against this. How do we do this properly? (We always want to depend on the same scala version build).

Reply:
`resourceGenerators` takes a list of `Task`s, so you should (in theory - aka, in sbt, with blood, spit and lots of looking at the internals of sbt's macros) be able to wire `products in (linkerPrivateLibrary, Compile)` up so it runs the tasks within the task graph.

For starters you'll have to statically define those (`products in (compiler, Compile)`, `products in (library, Compile)`, etc) and then pull out the underlying task value (`.taskValue`?) and then modify those results to finish defining your resource generator.  Look at the old "tuple syntax" code for some ideas.

Reply:
Thanks for the quick answer. That would work, but I'm not 100% sure what the difference with simply writing `val privateLibProducts = (products in (linkerPrivateLibrary, Compile)).value` would be.

I think the main challenge is that we need to inject different states into parts of the task graph.

For example, if we want to link the test suite in 2.13.x, we need to:

- Compile the SJS compiler for 2.13.x
- Compile the test suite for 2.13.x (using the SJS compiler)
- Compile the SJS compiler for 2.12.8
- Compile the linker library for 2.12.8 (using the SJS compiler)
- Compile the linker for 2.12.8 (using a plain compiler)
- Link the test suite artifacts with the linker.

So the main challenge I see here is that we need to compile the compiler twice. So we need to run these tasks twice under different states. Is that doable?

Reply:
> I'm not 100% sure what the difference with simply writing...

You're right - that's probably the same.

> So the main challenge I see here is that we need to compile the compiler twice.

You need different projects then. `products in (compiler213, Compile)`, etc.

Reply:
@sjrd what do you say about this? We could do this, but it would probably make the community build integration even more complicated :-/

Reply:
It looks like it's the only safe way, yes.

It's probably not more complicated for the community build than what we are already dealing with in #3805. If we create a `compilerForLinkerPrivateLib` and a `libraryForLinkerPrivateLib` in addition to the `linkerPrivateLib`, those three projects would have to be pinned to 2.12.8, just like the `ir` and `linker` projects.